File: python-sdks/examples/openai-sdk/decorators.py
"""Decorators — Use @workflow and @task to structure traces."""

import os
from dotenv import load_dotenv

load_dotenv(override=True)

from openai import OpenAI
from respan import Respan, workflow, task
from respan_instrumentation_openai import OpenAIInstrumentor

respan = Respan(instrumentations=[OpenAIInstrumentor()])

client = OpenAI(
    api_key=os.getenv("RESPAN_API_KEY"),
    base_url=os.getenv("RESPAN_BASE_URL", "https://api.respan.ai/api"),
)


@task(name="generate_outline")
def generate_outline(topic: str) -> str:
    response = client.chat.completions.create(
        model="gpt-4.1-nano",
        messages=[
            {"role": "system", "content": "Generate a 3-point outline. Be concise."},
            {"role": "user", "content": topic},
        ],

    )
    return response.choices[0].message.content


@task(name="write_draft")
def write_draft(outline: str) -> str:
    response = client.chat.completions.create(
        model="gpt-4.1-nano",
        messages=[
            {"role": "system", "content": "Write a short paragraph from this outline."},
            {"role": "user", "content": outline},
        ],

    )
    return response.choices[0].message.content


@workflow(name="content_pipeline")
def run(topic: str) -> str:
    outline = generate_outline(topic)
    print(f"Outline:\n{outline}\n")

    draft = write_draft(outline)
    print(f"Draft:\n{draft}")
    return draft


run("Benefits of open-source software")
respan.flush()

File: javascript-sdks/examples/openai-sdk/decorators.ts
/** Decorators — Use withWorkflow and withTask to structure traces. */

import "dotenv/config";
import OpenAI from "openai";
import { Respan } from "@respan/respan";
import { OpenAIInstrumentor } from "@respan/instrumentation-openai";

const respan = new Respan({
  apiKey: process.env.RESPAN_API_KEY,
  baseURL: process.env.RESPAN_BASE_URL,
  instrumentations: [new OpenAIInstrumentor()],
});
await respan.initialize();

const client = new OpenAI({
  apiKey: process.env.RESPAN_API_KEY,
  baseURL: process.env.RESPAN_BASE_URL,
});

async function generateOutline(topic: string): Promise<string> {
  return respan.withTask({ name: "generate_outline" }, async () => {
    const response = await client.chat.completions.create({
      model: "gpt-4.1-nano",
      messages: [
        { role: "system", content: "Generate a 3-point outline. Be concise." },
        { role: "user", content: topic },
      ],
    });
    return response.choices[0].message.content!;
  });
}

async function writeDraft(outline: string): Promise<string> {
  return respan.withTask({ name: "write_draft" }, async () => {
    const response = await client.chat.completions.create({
      model: "gpt-4.1-nano",
      messages: [
        { role: "system", content: "Write a short paragraph from this outline." },
        { role: "user", content: outline },
      ],
    });
    return response.choices[0].message.content!;
  });
}

async function run(topic: string) {
  return respan.withWorkflow({ name: "content_pipeline" }, async () => {
    const outline = await generateOutline(topic);
    console.log(`Outline:\n${outline}\n`);

    const draft = await writeDraft(outline);
    console.log(`Draft:\n${draft}`);
    return draft;
  });
}

await run("Benefits of open-source software");
